mansrc/load2db.1
'\" t
.\"     Title: load2db
.\"    Author: [see the "AUTHORS" section]
.\" Generator: Asciidoctor 1.5.5
.\"      Date: 
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "LOAD2DB" "1" "" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.LINKSTYLE blue R < >
.SH "NAME"
load2db \- load genome sequences and extrinsic evidence hints into a MySQL database
.SH "SYNOPSIS"
.sp
\fBload2db\fP [OPTIONS] \-\-species=\fISPECIES\fP \-\-dbaccess=\fIdbname,host,user,passwd\fP \fIinputfilename\fP
.SH "DESCRIPTION"
.sp
\fBload2db\fP is a tool to load genome sequences and extrinsic evidence hints into a MySQL database.
.br
This database must be created before calling \fBload2db\fP.
.br
When storing genomes/hints of multiple organisms call this program repeatedly for each one.
.SH "OPTIONS"
.sp
\fIinputfilename\fP refers to a genome file in FASTA format or a hints file in GFF format.
.SS "Mandatory options"
.sp
\fB\-s\fP, \fB\-\-species\fP=\fISPECIES\fP
.RS 4
SPECIES is a name for your species and the same identifier as is used in the treefile and alnfile parameters to augustus.
.RE
.sp
\fB\-d\fP, \fB\-\-dbaccess\fP=\fIdbname,host,user,passwd\fP
.RS 4
dbname,host,user,passwd are the name of the SQL database, the host name or IP, the database user and password
.RE
.SS "Additional options"
.sp
\fB\-c\fP, \fB\-\-chunksize\fP=\fIsize\fP (\(lA 1000000)
.RS 4
This option is only relevant when loading a sequence file.
.br
The sequences in the input genome are split into chunks of this size so
that subsequent retrievals of small sequence ranges do not require to read
the complete \- potentially much longer \- chromosome. Default Value: 50000
.RE
.sp
\fB\-h\fP, \fB\-\-help\fP
.RS 4
Produce help message.
.RE
.SH "EXAMPLE"
.sp
.if n \{\
.RS 4
.\}
.nf
  create database before calling load2db:
   $ mysql \-u root \-p
   > CREATE DATABASE vertebrates;
   > CREATE USER \(aqmyuser\(aq@\(aq%\(aq IDENTIFIED BY \(aqmypasswd\(aq; /* or whatever the password code is*/
   > GRANT ALL ON vertebrates.* TO \(aqmyuser\(aq@\(aq%\(aq IDENTIFIED BY \(aqmypasswd\(aq;

  load2db \-\-species=mouse \-\-dbaccess=vertebrates,localhost,myuser,mypasswd mouse.fa
  load2db \-\-species=mouse \-\-dbaccess=vertebrates,localhost,myuser,mypasswd mouse.hints.gff
  load2db \-\-species=human \-\-dbaccess=vertebrates,localhost,myuser,mypasswd human.fa
  load2db \-\-species=human \-\-dbaccess=vertebrates,localhost,myuser,mypasswd human.hints.gff
.fi
.if n \{\
.RE
.\}
.SH "AUTHORS"
.sp
AUGUSTUS was written by M. Stanke, O. Keller, S. König, L. Gerischer and L. Romoth.

HISTORY.TXT
     - remove bam2wig
     - compile with CGP and ZIPINPUT support by default
     - compile with SQLite and MySQL support for CGP (switch of by adding SQLite = false and MySQL = false to common.mk)
List of changes from version 3.3.2 to 3.3.3 (until Sep 13th, 2019)
     - docker container
     - new script tfix_in_frame_stop_codon_genes.py that replaces genes where spliced in-frame stop codons were predicted
     - new scripts: compare_masking.pl, merge_masking.pl
     - new program bamToWig.py as alternarive to bam2wig C++ binary
     - fix warnings on new GCC compiler (8.3)
     - introduced unit tests of C++ code (make test)
     - bugfixes (lex.cc, autoAug.pl)
     - DIAMOND as alternative to BLAST in aa2nonred.pl
List of changes from version 3.3.1 to 3.3.2 (until Oct 5th, 2018)
     - bugfixes in comparative augustus, utrrnaseq
     - new species Chiloscyllium punctatum (bamboo shark), Scyliorhinus torazame (cat shark), Rhincodon typus (whale shark)
     - updated comparative augustus (CGP) tutorial 
List of changes from version 3.3 to 3.3.1 (until May 8th, 2018)
     - new species pisaster (Pisaster ochraceus, ochre starfish)
     - bugfixes of sampling error in intron model and in check of
       frame-compatibility of CDSpart hints without frame
     - new auxiliary program utrrnaseq to find training UTRs
     - new script setStopCodonFreqs.pl
List of changes from version 3.2.3 to 3.3 (until July 11th, 2017)
     - new program ESPOCA to estimate selective pressure on codon alignments
     - gene finding on ancestral genomes is enabled
     - new default parameters for comparative gene prediction (CGP)
     - clade parameters training for CGP
     - compatibility to Ian Fiddes' Comparative Annotation Toolkit (CAT)
     - new scripts eval_dualdecomp.pl,
     - more tolerant tree parsing
     - bugfixes in augustus, joingenes, load2sqlitedb, transMap2hints.pl, splitMfasta.pl, intron2exex.pl,
       aln2wig
     - new functionality in homGeneMapping, joingenes
     - new species ciona, mnemiopsis_leidyi, nematostella_vectensis, strongylocentrotus_purpuratus
List of changes from version 3.2.2 to 3.2.3 (until December 6th, 2016)
     - bugfixes in joingenes, load2sqlitedb, transMap2hints.pl
     - added script gtf2bed.pl
     - CGP parameters for clade of 8 primates (cgp/log_reg_parameters_primates.cfg)
     - added option --clean to load2sqlitedb
     - now filter out truncated coding genes with CDS length <= 3
     - Sascha Steinbiss: remove dependency on PATH_MAX
List of changes from version 3.2.1 to 3.2.2 (until March 30th, 2016)
     - added extrinsic.M.RM.PB.cfg for PacBio CCS hints with GMAP
       (see http://bioinf.uni-greifswald.de/bioinf/wiki/pmwiki.php?n=Augustus.PacBioGMAP )
     - standardization of Makefile
     - improvements in auxiliary program joingenes
     - added missing program load2sqlitedb
     - comparative gene predictions (CGP):
         - hints with source=M (e.g. existing annotations) are fixed
	 - new standard parameters for exon_gain, exon_loss and others
     - bugfixes, in particular a double free or corruption bug in PPX that occured with BUSCO
     - new species Volvox carteri (volvox)
List of changes from version 3.2 to 3.2.1 (until October 23rd, 2015)
     - comparative gene prediction (CGP): make the parameters from the GCB proceedings paper match again
     - new CGP default parameters (tested on 5 clades)
     - CGP version penalizes very long introns equivalently to a geometric distribution
     - code now under artistic license only
List of changes from version 3.1 to 3.2 (until September 25th, 2015)
     - comparative gene prediction (CGP option) has sensible default parameters of the logistic regression model
     - new tool hal2maf_split.pl for splitting an alignment for parallel comparative prediction
     - bugfixes to augustus (compiler warnings on MaxOS)
     - improvements to prediction of noncoding genes
     - homGeneMapping can determine homologous sets of transcripts
     - bugfixes in joingenes and bam2hints (rare segmentation fault)
     - made load2sqlitedb more robust, new option to skip index building when loading many species
     - new tutorial that includes iterative training
     - new option to shift coordinates in hint input when sequence window is cut out
     - adapted cegma2gff.pl to cegma version 2.5
List of changes from version 3.0.3 to 3.1 (until April 29th, 2015)
     - prediction of noncoding genes (ncRNA) with option --nc=on, requires input of hints, e.g. from RNA-Seq,
       predicts coding genes (with UTRs) and noncoding genes simultaneously
     - speed improvement with --UTR=1 (by 83% on human without sampling, by 18% on fly with sampling)
     - softmasking support: --softmasking=1 is a little more accurate and much faster on human
     - small overall accuracy improvement
     - prevent the prediction of spliced in-frame stop codons (for SHORT or HINTED introns only)
     - new species: rice, Danio rerio (zebrafish), s_aureus (Staphylococcus aureus),
       camponotus_floridanus (ant) contributed by Shishir K Gupta,
       sulfolobus_solfataricus (archaeon), Parasteatoda tepidariorum (house spider), Apis dorsata (from Francisco Camara Ferreira)
     - update to parameters of Thermoanaerobacter tencongensis, E.coli, Yarrowia lipolytica, schistosoma2 (Schistosoma mansoni)
     - remove dependency on libboost-filesystems, boost is now completely optional in single-species mode
     - new auxiliary program 'homGeneMapping' for identification of homologous genes in a set of gene predictions
     - new auxiliary program 'joingenes' for combining gene sets
     - bugfixes in load2db, length distribution skew of bacteria, comparative mode (findGeneRanges,
       Alignment::mergeable, memory leak), some ass hints were previously not used although possible
     - decreased minintronlen default of bam2hints to 32 as the old threshold tended to be too large
     - computing of omega = dN/dS more efficiently

src/load2db.cc
/*
 * load2db.cc
 *
 * License: Artistic License, see file LICENSE.TXT or 
 *          https://opensource.org/licenses/artistic-license-1.0
 * 
 * Description: Load the sequences from a flat file in fasta format into a database.
 */

// Project includes
#include "fasta.hh"
#include "projectio.hh"
#include "hints.hh"

// standard C/C++ includes
#include <string>
#include <iostream>
#include <fstream>
#include <getopt.h>     /* for getopt_long; standard getopt is in unistd.h */
#include <stdlib.h>     /* for exit() */
#include <mysql++.h>
#include <exception>
#include <ssqls.h>

//#include <boost/iostreams/filtering_stream.hpp>
//#include <boost/iostreams/filter/gzip.hpp>
//#include <boost/iostreams/copy.hpp>

using namespace std;
//using boost::iostreams::filtering_istream;
//using boost::iostreams::gzip_decompressor;

//sql_create_6(genomes,1,6,int,seqid,string,sequence,string,seqname,int,start,int,end,string,species)

int chunksize = 50000;
mysqlpp::Connection con;

void printUsage();
void connectDB(string dbaccess);
void createTableGenomes();
void createTableSpeciesnames();
void createTableSeqnames();
void createTableHints();
void createTableFeatureTypes();
int getSpeciesID(string species);
int getSeqNr(char* seqname,int speciesid);
int insertSeqName(char* seqname,int speciesid);
int insertSeq(string sequence, char* name, int length, int speciesid);
void insertHint(Feature &f, string species);
void checkConsistency(); // checks consistency of the 'hints' data with the 'genomes' data
void removeDuplicates();

/*
 * main
 */
int main( int argc, char* argv[] ){
    int c;
    int help = 0;
    string species;
    string dbaccess;
    string filename;
    static struct option long_options[] = {
        {"species", 1, 0, 's'},
        {"dbaccess", 1, 0, 'd'},
        {"help", 0, 0, 'h'},
	{"chunksize", 1, 0, 'c'},
        {NULL, 0, NULL, 0}
    };
    int option_index = 0;
    while ((c = getopt_long(argc, argv, "s:d:hc:", long_options, &option_index)) != -1) {
        switch (c) {
        case 's':
	    species = optarg;
            break;
        case 'd':
	    dbaccess = optarg;
            break;
        case 'h':
            help = 1;
            break;
        case 'c':
	    chunksize = atoi(optarg);
            break;
        default:
            break;
        }
    }
    if (optind < argc-1) {
        cerr << "More than one option without name: ";
        while (optind < argc)
            cerr << " " << argv[optind++];
	cerr << endl << "Only the sequence/hints file name does not require a parameter name." << endl;
	if (!help)
	    printUsage();
	exit(1);
    }
    if (optind == argc-1)
	filename = argv[optind];
    else {
	cerr << "Missing sequence/hints file name." << endl;
	printUsage();
	exit(1);
    }
    if (help) {
	printUsage();
	exit(1);
    }
    if (species.empty()){
	cerr << "Missing species name. Required parameter." << endl;
	printUsage();
	exit(1);
    }
    if (dbaccess.empty()){
	cerr << "Missing database access info. dbaccess is a required parameter." << endl;
	printUsage();
	exit(1);
    }
    if (chunksize < 2){
	cerr << "Chunksize too small (" << chunksize << "). Should be roughly in the oder of a gene's length." << endl;
	printUsage();
	exit(1);
    }
    if (chunksize > 1000000){
	cerr << "Chunksize too big (" << chunksize << "). " << endl;
	printUsage();
	exit(1);
    }

    try {
	connectDB(dbaccess);
    } catch (const char *m) {
	cerr << "Database connection error:" << endl << m << endl;
	exit(1);
    }
    try {
	createTableSpeciesnames();
	createTableSeqnames();

	ifstream ifstrm;
	ifstrm.open(filename.c_str());
	if( !ifstrm )
	    throw string("Could not open input file \"") + filename + "\"!";

	/*filtering_istream zin;
	try {  
	    zin.push(gzip_decompressor());
	    zin.push(ifstrm);
	    zin.peek();
	    if (!zin)
		throw("Could not read first character assuming gzip format.");
	    cout << "Looks like " << filename << " is a gzip file. Deflating..." << endl;
        } catch (...) { // boost::iostreams::gzip_error& 
	    // not a gzip file or ill-formatted
	    zin.reset();
	    ifstrm.seekg(0);
	    zin.push(ifstrm);
	    }*/
	
	//if(isFasta(zin)){
	if(isFasta(ifstrm)){
	    cout << "Looks like " << filename << " is in fasta format." << endl;
	    createTableGenomes();
	    char *sequence = NULL, *name = NULL;
	    int length = 0, seqCount = 0, chunkCount = 0;
	    unsigned int lenCount = 0;
	    //readOneFastaSeq(zin, sequence, name, length);
	    readOneFastaSeq(ifstrm, sequence, name, length);
	    int speciesid = getSpeciesID(species);
	    while (sequence){
		chunkCount += insertSeq(sequence, name, length, speciesid);
		seqCount++;
		lenCount += length;
		delete sequence;
		delete name;
		sequence = name = NULL;
		//readOneFastaSeq(zin, sequence, name, length);
		readOneFastaSeq(ifstrm, sequence, name, length);
	    }
	    if (seqCount > 0)
		cout << "Inserted " << chunkCount << " chunks of " << seqCount << " sequences (total length "
		     << lenCount << " bp)." << endl;
	    else
		cout << "No sequences found. Nothing inserted into database." << endl;
	    
	}
	else if(isGFF(ifstrm)){
	    cout << "Looks like " << filename << " is in gff format." << endl;
	    createTableHints();
	    createTableFeatureTypes();
	    map<string,int> hintCount; // stores for each sequence the number of hints inserted into the database
	    while(ifstrm){
		Feature f;
		try{
		    ifstrm >> f >> comment >> ws;
		    if(f.type != -1){
			insertHint(f,species);
			hintCount[f.seqname]++;
		    }
		} catch (ProjectError e){}
	    }
	    if(!hintCount.empty()){
		cout << "inserted" << endl; 
		for(map<string,int>::iterator it= hintCount.begin(); it != hintCount.end(); it++){
		    cout << it->second << " hints for " << it->first << endl;
		}
		//removeDuplicates(); dublicastes can only be removed when multiplicity is updated
	    }
	    else{
		cout << "No hints found. Nothing inserted into database." << endl;
	    }
	    checkConsistency();
	}
	else{
	    cout << filename << " is neither in gff nor in fasta format." << endl;
	}
	ifstrm.close();   
    } catch( string err ){
        cerr << "\n" <<  argv[0] << ": ERROR\n\t" << err << "\n\n";
	exit(1);
    } 
}

void printUsage(){
    cout << "usage:\n\
load2db [parameters] --species=SPECIES --dbaccess=dbname,host,user,passwd  inputfilename\n\
\n\
inputfilename refers to a genome file in FASTA format or a hints file in GFF format\n\
SPECIES is the same identifier as is used in the treefile and alnfile parameters to augustus.\n\
\n\
dbname,host,user,passwd are the name of the SQL database, the host name or IP, the database user and password\n\
When storing genomes/hints of multiple organisms call this program repeatedly for each one.\n\
A single table with the structure\n\
\
is created.\n\
\n\
parameters:\n\
--help        print this usage info\n\
--chunksize   this option is only relevant when loading a sequence file\n\
              the sequences in the input genome are split into chunks of this size so\n\
              that subsequent retrievals of small sequence ranges do not require to read\n\
              the complete - potentially much longer - chromosome. (<= 1000000, default " << chunksize << ")\n\
\n\
example:\n\
     load2db --species=chicken --dbaccess=birds,localhost,mario,dF$n.E chickengenome.fa\n\
     load2db --species=chicken --dbaccess=birds,localhost,mario,dF$n.E chickenhints.gff\n\
\n\
Example code for database creation before calling load2db:\n\
mysql -u root -p\n\
create database birds;\n\
select password('dF$n.E');\n\
create user `mario`@`%` identified by password '*72E7F76393830492EB3C58AD730188708BD72DE1'; /* or whatever the password code is*/\n\
grant all privileges on birds.* to mario@'%';\n";
}

void connectDB(string dbaccess){
    string db_name, host, user, passwd;

    string::size_type start=0, end;
    end = dbaccess.find(','); // string 'dbaccess' is delimited by ','
    if (end == string::npos)
	throw ("Database name missing in dbaccess.");
    db_name = dbaccess.substr(start, end-start);
    start = end + 1;
    end = dbaccess.find(',', start);
    if (end == string::npos)
	throw ("Host name missing in dbaccess.");
    host = dbaccess.substr(start, end-start);
    start = end + 1;
    end = dbaccess.find(',', start);
    if (end == string::npos)
	throw ("User name missing in dbaccess.");
    user = dbaccess.substr(start, end-start);
    start = end + 1;
    end = dbaccess.find(',', start);
    if (end == string::npos)
	passwd = dbaccess.substr(start);
    else 
	passwd = dbaccess.substr(start, end-start); // in case dbaccess also includes the port number
    
    try {
	cout << "Trying to connect to database " << db_name << " on server "
	     << host << " as user " << user << " using password " << passwd << " ..." << endl;
	con.connect(db_name.c_str(), host.c_str(), user.c_str(), passwd.c_str());
    } catch(const mysqlpp::BadQuery& e){
	cout << "Connection error: " << e.what() << endl;
    }
}

/*
 * create table `genomes` if it does not already exist
 */
void createTableGenomes(){
    mysqlpp::Query query = con.query("CREATE TABLE IF NOT EXISTS genomes (\
        seqid int(10) unsigned NOT NULL AUTO_INCREMENT,			\
        dnaseq longtext NOT NULL,					\
        seqnr int(10) unsigned,					\
        start int(9) unsigned NOT NULL,\
        end int(9) unsigned NOT NULL,\
        speciesid int(10) unsigned,\
        PRIMARY KEY (seqid),\
        KEY region (speciesid,seqnr,start,end),\
        FOREIGN KEY (speciesid,seqnr) REFERENCES seqnames(speciesid,seqnr)\
      ) ENGINE=MyISAM DEFAULT CHARSET=latin1 MAX_ROWS=10000000 AVG_ROW_LENGTH=50000;");
   query.execute();
}
/*
 * create table 'speciesnames' if it does not already exist
 */
void createTableSpeciesnames(){
    mysqlpp::Query query = con.query("CREATE TABLE IF NOT EXISTS speciesnames (\
        speciesid int(10) unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY,\
        speciesname varchar(50) UNIQUE KEY\
      ) ENGINE=MyISAM DEFAULT CHARSET=latin1 MAX_ROWS=1000000 AVG_ROW_LENGTH=50000;");
    query.execute();
}
/*                                                                                                                                                                                                                
 * create table 'seqnames' if it does not already exist                                                                                                                                                        
 */
void createTableSeqnames(){
    mysqlpp::Query query = con.query("CREATE TABLE IF NOT EXISTS seqnames (\
        seqnr int(10) unsigned AUTO_INCREMENT,\
        speciesid int(10) unsigned REFERENCES speciesnames(speciesid),\
        seqname varchar(50),\
        PRIMARY KEY(speciesid,seqnr),\
        UNIQUE KEY(speciesid,seqname)\
      ) ENGINE=MyISAM DEFAULT CHARSET=latin1 MAX_ROWS=1000000 AVG_ROW_LENGTH=50000;");
    query.execute();
}
/*
 * create table 'hints' if it does not already exist
 */
void createTableHints(){
    mysqlpp::Query query = con.query("CREATE TABLE IF NOT EXISTS hints (\
        hintid int(10) unsigned AUTO_INCREMENT PRIMARY KEY,\
        speciesid int(10) unsigned,\
        seqnr int(10) unsigned,\
        source varchar(50),\
        start int(9) unsigned NOT NULL,\
        end int(9) unsigned NOT NULL,\
        score float NOT NULL DEFAULT 0.0,\
        type tinyint unsigned NOT NULL,\
        strand enum('+','-','.') NOT NULL DEFAULT '.',\
        frame enum ('0','1','2','.') NOT NULL DEFAULT '.',\
        priority smallint NOT NULL DEFAULT -1,\
        grp varchar(100) DEFAULT '',\
        mult smallint unsigned DEFAULT 1,\
        esource varchar(10) NOT NULL,\
        KEY region (speciesid,seqnr,start,end),\
        FOREIGN KEY (speciesid,seqnr) REFERENCES seqnames(speciesid,seqnr)\
 ) ENGINE=MyISAM DEFAULT CHARSET=latin1 MAX_ROWS=1000000000 AVG_ROW_LENGTH=50000;");
    query.execute();
}

void createTableFeatureTypes(){
    mysqlpp::Query query = con.query("CREATE TABLE IF NOT EXISTS featuretypes (\
        typeid tinyint unsigned PRIMARY KEY,\
        typename enum('start','stop','ass','dss','tss','tts','exonpart','exon','intronpart','intron',\
        'irpart','CDS','CDSpart','UTR','UTRpart','nonexonpart','genicpart') NOT NULL\
 ) ENGINE=MyISAM DEFAULT CHARSET=latin1 MAX_ROWS=1000000 AVG_ROW_LENGTH=50000;");
    query.execute();
    for(int i=0; i < 17; i++){
	query << "INSERT IGNORE INTO featuretypes VALUES(" << i << "," << (i+1) << ")";
	//cout << "Executing" << endl << query.str() << endl;
	query.execute();
    }
}

/*
 * insert one sequence into database
 * return the number of chunks that sequence was cut into
 * start and end are 0-based and inclusive.
 */
int insertSeq(string sequence, char *name, int length, int speciesid){
    int chunks = 0;
    int start = 0, end;
    int seqnr = insertSeqName(name,speciesid);
    while (start < length) {
	end = (start + chunksize < length)? (start + chunksize - 1) : length - 1;
	mysqlpp::Query query = con.query();
	query << "INSERT INTO genomes (dnaseq,seqnr,speciesid,start,end) VALUES(\""
	      << sequence.substr(start, end-start+1) << "\"," << seqnr << "," 
	      << speciesid << "," << start << "," << end << ")";
	//cout << "Executing" << endl << query.str() << endl;
	query.execute();
	chunks++;
	start += chunksize;
    }
    return chunks;
}
/*
 * insert one hint into database
 * start and end are 0-based and inclusive
 */
void insertHint(Feature& f, string species){
    mysqlpp::Query query = con.query();
    int speciesid = getSpeciesID(species);
    int seqnr = getSeqNr(&(f.seqname[0]),speciesid);
    string attributes = "";
    string values = "";
    switch (f.strand) {
        case plusstrand: attributes+="strand,"; values+=("\"+\","); break;
        case minusstrand: attributes+="strand,"; values+=("\"-\","); break;
        default : break;
    }
    if (f.frame != -1){
	attributes+="frame,";
	values+=("\"" + itoa(f.frame) + "\",");
    }
    if (f.groupname != ""){
	attributes+="grp,";
	values+=("\"" + f.groupname + "\",");
    }
    if (f.mult > 1){
	attributes+="mult,";
	values+=(itoa(f.mult) + ",");
    }
    if (f.priority >= 0){
	attributes+="priority,";
	values+=(itoa(f.priority) + ",");
    }
    query << "INSERT INTO hints (speciesid,seqnr,source,start,end,score,type," << attributes << "esource) VALUES ("
	  << speciesid << "," << seqnr <<  ",\"" << f.source << "\"," << f.start << "," << f.end << "," << f.score << ","
          << f.type << "," << values << "\""<< f.esource << "\")";
    //cout << "Executing" << endl << query.str() << endl;
    query.execute();
  
}
/*
 * returns the speciesid for a given species name
 * if the species name is not in the database it is inserted
 */
int getSpeciesID(string species){
    mysqlpp::Query query = con.query();
    query << "SELECT speciesid FROM speciesnames WHERE speciesname='" <<species<<"'";;
    //cout << "Executing" << endl << query.str() << endl;
    mysqlpp::StoreQueryResult res = query.store();
    if(!res.empty()){
	return res[0]["speciesid"];
    }
    else{
	// insert species name
	query << "INSERT INTO speciesnames (speciesname) VALUES (\"" << species << "\")";
	//cout << "Executing" << endl << query.str() << endl;
	query.execute();
	return query.insert_id();
    }
}
/*
 * returns the seqnr for a given sequence name and species id
 * it the sequence name is not in the database, it will be inserted
 */
int getSeqNr(char* seqname,int speciesid){
    mysqlpp::Query query = con.query();
    query << "SELECT seqnr FROM seqnames WHERE seqname='" << seqname << "' AND speciesid=" << speciesid;
    // cout << "Executing" << endl << query.str() << endl;
    mysqlpp::StoreQueryResult res = query.store();
    if (!res.empty()){
        return res[0]["seqnr"];
    } else{
        query << "INSERT INTO seqnames (speciesid,seqname) VALUES (" << speciesid << ",\"" << seqname << "\")";
	// cout << "Executing" << endl << query.str() << endl;
        query.execute();
        return query.insert_id();
    }
}
int insertSeqName(char* seqname,int speciesid){
    try{
	mysqlpp::Query query = con.query();
	query << "INSERT INTO seqnames (speciesid,seqname) VALUES (" << speciesid << ",\"" << seqname << "\")";
	//cout << "Executing" << endl << query.str() << endl;
	query.execute();
	return query.insert_id();
    } catch(const mysqlpp::BadQuery& e){
        cerr <<"sequence "<< seqname <<" already exists in database" << endl;
	cerr <<"Please delete sequence, before reloading"<<endl;
	exit(1);
    }
}

void checkConsistency(){
    mysqlpp::Query query = con.query("SELECT DISTINCT speciesname,seqname FROM hints as h,\
        speciesnames as s,seqnames as n where h.speciesid=s.speciesid AND\
        s.speciesid=n.speciesid AND h.seqnr=n.seqnr AND NOT EXISTS\
        (SELECT DISTINCT speciesid,seqnr FROM genomes as g WHERE g.speciesid=h.speciesid AND g.seqnr=h.seqnr);");
    mysqlpp::StoreQueryResult res = query.store();
    if(!res.empty()){
	cout << "WARNING: the following sequences are not in the database although we have hints for them" << endl;
	for(size_t i=0;i<res.num_rows();i++){
	    cerr<<res[i]["seqname"]<<" (species "<<res[i]["speciesname"]<<")"<<endl;
	}
    }
}
void removeDuplicates(){
    cout << "removing duplicate hints"<< endl;
    mysqlpp::Query query = con.query(" ALTER IGNORE TABLE hints ADD UNIQUE\
        (speciesid,seqnr,source,start,end,score,type,strand,frame, priority,grp,mult,esource);");
    mysqlpp::SimpleResult res = query.execute();
    cout << res.info() << endl;
}
